### ISSUE TITLE
Change Request: Support multiple error locations for a reported problem

### ISSUE THREAD

**Issue by DMartens** [eslint/eslint#20176]
### ESLint version

latest

### What problem do you want to solve?

# Problem
ESLint currently does not support multiple error locations for a single reported problem.
Providing the related locations helps users understand and resolve the lint problem.
For example when a rule reports a mutation of a supposed constant the location of the lint message could be the assignment and
the related location could be the initial definition.

Currently a workaround for providing the additional information is to use a long / multiline message.
But it is hard to format the same message for different sizes (terminal and editors).
Further editors could show the related locations only on hover and allow to jump there.

# Examples
|Rule|Lint Message Location|Additional Location|
|------|---------------|---|
|`no-const-assign`|Assignment| Initial variable declaration|
|`no-dupe-keys`|Duplicated property key|First property key of the same name|
|`grouped-accessor-pairs`|Latter accessor|Corresponding accessor|
|`no-loop-func`|Assignment|Outer loop|
|`preserve-caught-error`|`throw`|Catch binding|

# Other Tools
- General: [Language Server Protocol](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticRelatedInformation), [tsserver](https://github.com/microsoft/TypeScript/blob/main/src/server/protocol.ts)
- Linter: [clippy](https://doc.rust-lang.org/stable/clippy/development/emitting_lints.html), [go-analyzers](https://pkg.go.dev/golang.org/x/tools/go/analysis#Diagnostic)
- Compiler: LLVM

### What do you think is the correct solution?

This could be solved by providing an additional property `relations` in the report descriptor of `context.report`.
An example for `no-const-assign`:
```javascript
context.report({
    // ... Other properties
    relations: {
        declaration: {
           description: 'Defined as a constant here',
           loc:         declaration.loc,
        },
    },
});
```

This change would be backwards compatible as additional properties are allowed for report descriptors.
But the complete design of this requires an RFC, which I would write if this proposal is accepted.

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

This issue is inspired by / continuation of the discussion #20043.

**Comment by JoshuaKGoldberg** [comment]
Ha, I'd planned on requesting this feature too - thanks for getting it started 😄 .

+1 to the idea of adding in secondary / "relations" ranges on the report. It's useful and supported by LSPs now. 

In case it's helpful, quoting https://github.com/eslint/eslint/discussions/20043#discussioncomment-14568152's references to other linters' equivalent concepts:

* Biome example: [playground of its `noArrayIndexKey` rule](https://biomejs.dev/playground/?lintRules=noArrayIndexKey&code=CgBzAG8AbQBlAHQAaABpAG4AZwAuAGYAbwByAEUAYQBjAGgAKAAoAEUAbABlAG0AZQBuAHQALAAgAGkAbgBkAGUAeAApACAAPQA%2BACAAewAKACAAIAAgACAAPABDAG8AbQBwAG8AbgBlAG4AdAAgAGsAZQB5AD0AewBpAG4AZABlAHgAfQAgAD4AZgBvAG8APAAvAEMAbwBtAHAAbwBuAGUAbgB0AD4ACgB9ACkAOwA%3D&ruleDomains.react=recommended) <- [source code of detail range](https://github.com/biomejs/biome/blob/fa4fcd355f8bce46b83a60fc2a874bfa93c2ce4f/crates/biome_js_analyze/src/lint/suspicious/no_array_index_key.rs#L235-L237)
* Oxlint example: [playground of its `eslint/for-direction` rule](https://playground.oxc.rs/#eNptkk1v2zAMhv+KodNWJGg7YMCQbad1AQrsCzvs5AsrM546mTJIOs1g+L+PciNlh54kvRSplw81O+927vqqQYmBtDkk3naB0WtItGtah8yJW9dcXbfUkoWbV0fgJjQfm5v3tnxobvO63b5u5paWltzGJbebHU+Ul1zV7ZQn3DjLHkAV2e0OEOV/5f6iKQNJDlQlSIqg2N2hj8CQrUkNPv0OijKCxyoNQH28HH0aRka5pIhPo4WfTcnf4SHFcvKH/nxt2bgRWLLZ2eFJkcTeNVgqJ+sRYkxPP1Enpu+TSuhwP9FKrZTKbyIf8Qew5RZ1zfv17p6Ug1X0NSBotjX4zxn4WTUPmd/qwfaVzLMpBe7R4DqUNze3b83UJHiHh0C4T/wpgsg+YOzqE3gakcOApBCNZTKU9SmLDkGr+hUVOlCoPl6ewezE+hitG2QCo1jQ+dRhj+sPoDzjCvhRumRf7gyDKOlaqygR+3J3rVIml9svY7XO0/gFj1iL/kEcv8GAZcRrqhFOcW+0c8YR+SFJ+RO5H3q0T55DdbeY3K34slx3S9Yz5tLT8g8PGCNC) <- [source code of label range](https://github.com/oxc-project/oxc/blob/bdf901033e424a894e81e248b6c0373d58495825/crates/oxc_linter/src/rules/eslint/for_direction.rs#L22-L25)

Additional +1 to an RFC.

**Comment by kecrily** [comment]
Really like this feature. The RFC should also include the new default formatter style, though it will likely resemble other tools offering this feature. Just a note
